sympy/stats/stochastic_process_types.py

```diff
@@ -676,31 +766,761 @@ def _transient2absorbing(self):
+        - A - the submatrix from recurrent states to recurrent states
+        - B - the submatrix from transient to recurrent states
+        - C - the submatrix from transient to transient states
+        - 0 - the submatrix of zeros for recurrent to transient states
+
+        Returns
+        =======
+        (states, A, B, C)
+
+            ``states`` - a list of state names with the first being
+            the recurrent states and the last being
+            the transient states in the order
+            of the row names of A and then the row names of C.
+            ``A`` - the submatrix from recurrent states to recurrent states.
+            ``B`` - the submatrix from transient to recurrent states.
+            ``C`` - the submatrix from transient to transient states.
+
+        Examples
+        ========
+        >>> from sympy.stats import DiscreteMarkovChain
+        >>> from sympy import Matrix, S
+
+        You can decompose this matrix for example
+
+        >>> T = Matrix([[S(1)/2, S(1)/2, 0, 0, 0],
+        ...             [S(2)/5, S(1)/5, S(2)/5, 0, 0],
+        ...             [0, 0, 1, 0, 0],
+        ...             [0, 0, S(1)/2, S(1)/2, 0],
+        ...             [S(1)/2, 0, 0, 0, S(1)/2]])
+        >>> X = DiscreteMarkovChain('X', trans_probs=T)
+        >>> X.decompose()
+        ([2, 0, 1, 3, 4], Matrix([[1]]), Matrix([
+        [  0],
+        [2/5],
+        [1/2],
+        [  0]]), Matrix([
+        [1/2, 1/2,   0,   0],
+        [2/5, 1/5,   0,   0],
+        [  0,   0, 1/2,   0],
+        [1/2,   0,   0, 1/2]]))
+
+        This means that state 2 is the only absorbing state
+        (since A is a 1x1 matrix). B is a 4x1 matrix since
+        the 4 remaining transient states all merge into reccurent
+        state 2. And C is the 4x4 matrix that shows how the
+        transient states 0, 1, 3, 4 all interact.
+
+        See Also
+        ========
+        sympy.stats.DiscreteMarkovChain.canonical_form
+
+        References
+        ==========
+        .. [1] https://en.wikipedia.org/wiki/Absorbing_Markov_chain
+        .. [2] http://people.brandeis.edu/~igusa/Math56aS08/Math56a_S08_notes015.pdf
+
+        """
+        trans_probs = self.transition_probabilities
+        if trans_probs is None:
+            return None
+        n = self.num_states
+
+        if isinstance(trans_probs, MatrixSymbol) or isinstance(n, Symbol):
+            raise NotImplementedError("The transient and recurrent states cannot be determined.")
+
+        classes = self.communication_classes()
+        r_states = []
+        t_states = []
+        for c in classes:
+            if c.recurrent:
+                r_states += c.states
+            else:
+                t_states += c.states
+
+        states = r_states + t_states
+        A = zeros(rows=len(r_states), cols=len(r_states))
+        for row in range(len(r_states)):
+            for col in range(len(r_states)):
+                A[row, col] = trans_probs[states[row], states[col]]
```

Comment:
Rather than using zeros and loops there is a shorthand for this sort of thing like:
```
In [2]: Matrix(4, 4, lambda i, j: i + j)                                                                                                                      
Out[2]: 
⎡0  1  2  3⎤
⎢          ⎥
⎢1  2  3  4⎥
⎢          ⎥
⎢2  3  4  5⎥
⎢          ⎥
⎣3  4  5  6⎦
```